duckduckgo/macos-browser
# Address bar refactoring

Task/Issue URL: https://app.asana.com/0/1177771139624306/1205421610519119/f
BSK PR: https://github.com/duckduckgo/BrowserServicesKit/pull/499
iOS PR: https://github.com/duckduckgo/iOS/pull/1989

**Description**:
- refactored String NSRange helpers to use designated functions instead of UTF16 view
- refactored AddressBarTextField for better structure and readability, added comments
- added Undo Manager support (cmd+z)
- Fixed opt+delete word forward deletion corrupting suffix
- Fixed opt+arrows by-word navigation
- Hopefully fixed a crash

**Steps to test this PR**:
1. Test address bar input, autocompletion selection for search terms, urls and bookmarks, performing search and editing url
2. Test opt+arrows for by-word navigation, test opt+backspace, opt+delete(fn+backspace) by word deletion
3. Test Undo/Redo for typing, entering words or separators, autocompleted suggestions and choosing autocompletion
4. Switching between tabs should keep the value and reset the Undo Manger
5. Validate testing steps from subtasks of https://app.asana.com/0/0/1205449500218048/f



---
###### Internal references:
[Pull Request Review Checklist](https://app.asana.com/0/1202500774821704/1203764234894239/f)
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
[Pull Request Documentation](https://app.asana.com/0/1202500774821704/1204012835277482/f)
